cirq-core/cirq/protocols/json_serialization.py

```diff
@@ -221,10 +220,22 @@ class CirqEncoder(json.JSONEncoder):
        See https://github.com/quantumlib/Cirq/issues/2014
     """
 
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self._memo: dict[Any, dict] = {}
+
     def default(self, o):
         # Object with custom method?
         if hasattr(o, '_json_dict_'):
-            return _json_dict_with_cirq_type(o)
+            json_dict = _json_dict_with_cirq_type(o)
+            if isinstance(o, SerializableByKey):
```

Comment:
It's true that `SerializableByKey` doesn't add any methods, but it still serves as a marker for types that want to opt in to this "deduplication" during serialization. One this that is being added here is that we now require that `SerializableByKey` classes must be hashable, which is not necessarily the case with all classes that implement `SupportsJSON`. If we want to change the opt-in mechanism that would be fine with me; I do think the _name_ `SerializableByKey` is a bit misleading (and has been since https://github.com/quantumlib/Cirq/pull/3673 when the `_serialization_key_` method was removed) so at very least we might consider changing the name.

Reply:
Good call. Renaming feels like it would be a headache to me, but given that this changes the serialization format anyways I guess now would be the time for it.

Happy to review any changes from this here or as a separate PR.

Reply:
Since the `SerializableByKey` name itself no longer appears in the serialized format, renaming will be a source-only change and would not affect compatibility of deserializing stored data, so that should be low risk. Nevertheless, since naming always invites bikeshedding, I'm inclined to defer that to a future PR, so I'll leave this as-is for now :-)